x/programs/wasmlanche/src/memory.rs

```diff
@@ -13,22 +13,106 @@ use alloc::{
     alloc::{alloc as allocate, dealloc as deallocate, handle_alloc_error, Layout},
     vec::Vec,
 };
-use core::{
-    cell::{LazyCell, RefCell},
-    mem::ManuallyDrop,
-    ops::Deref,
-    slice,
-};
-use hashbrown::HashMap;
+use core::{mem::ManuallyDrop, ops::Deref, slice};
+
+/// Safety:
+/// Only safe to use in a single-threaded environment
+mod allocations {
+    use cfg_if::cfg_if;
+    use core::cell::RefCell;
+    use hashbrown::HashMap;
+
+    type LenMap = HashMap<*const u8, usize>;
```

Comment:
That wouldn't work when allocating. You want the HostPtr to be returned, but it would be owned by the map. HostPtr isn't Clone or Copy, but *const u8 is. We don't want to make HostPtr Clone or Copy either as HostPtr's free on Drop.